src/main/java/Tree/BinarySearchTree.java

```diff
@@ -4,137 +4,179 @@
+    BinaryNode<T> root;
 
-    public BinarySearchTree() { current = null; }
+    public BinarySearchTree() { current = root = null; }
 
-    public BinarySearchTree(int key, T value) { current = new BinaryNode<T>(key, value); }
+    public BinarySearchTree(int key, T value) {
+        current = new BinaryNode<T>(key, value);
+        root = current;
+    }
+
+    public void setLeft(BinaryNode<T> leftChild, BinaryNode<T> parent) {
+        assert parent == null : "Parent node must not be null";
+        parent.setLeftNode(leftChild);
+        leftChild.setParentNode(parent);
+    }
+
+    public void setRight(BinaryNode<T> rightChild, BinaryNode<T> parent) {
+        assert parent == null : "Parent node must not be null";
+        parent.setRightNode(rightChild);
+        rightChild.setParentNode(parent);
+    }
 
     public boolean contains(int key) {
-        BinaryNode<T> original_position = current;
         Boolean result = false;
 
-        if (current == null) return false;
-
         while (current != null) {
             if (current.getKey() == key) {
                 result = true;
                 break;
-            } else if ((Integer) current.getKey() < (Integer) key) {
+            } else if (current.getKey() < key) {
                 moveToRightNode();
-            } else if ((Integer) current.getKey() > (Integer) key) {
+            } else if (current.getKey() > key) {
                 moveToLeftNode();
             }
         }
 
-        current = original_position;
+        current = root;
         return result;
     }
 
-    public void insert(int insert_key, T value) {
-        BinaryNode<T> original_position = current;
-        BinaryNode<T> node = new BinaryNode<T>(insert_key, value);
-
-        if (current == null) { current = node; }
+    public T get_value(int key) {
+        T result = null;
 
-        if (current.getKey() < insert_key) {
-            if (current.getRightNode() == null) {
-                setRightChild(node.getKey(), value);
-            } else {
+        while (current != null) {
+            if (current.getKey() == key) {
+                result = current.getValue();
+                break;
+            } else if (current.getKey() < key) {
                 moveToRightNode();
-                insert(insert_key, value);
-            }
-        } else if (current.getKey() > insert_key) {
-            if (current.getLeftNode() == null) {
-                setLeftChild(node.getKey(), value);
-            } else {
+            } else if (current.getKey() > key) {
                 moveToLeftNode();
-                insert(insert_key, value);
             }
         }
-        current = original_position;
-    }
 
-    public void delete(int delete_key) {
-        BinaryNode<T> original_position = current;
+        current = root;
+        return result;
+    }
 
-        if (delete_key == 0) return;
+    protected BinaryNode<T> get_node (int key) {
```

Comment:
method and variable should follow lowerCamelCase
https://google.github.io/styleguide/javaguide.html#s5.2.3-method-names